Cluster network by shapes (NUTS0, NUTS1, NUTS2, NUTS3, and ADM1)

bobbyxng opened the pull request:
**Closes https://github.com/PyPSA/pypsa-eur/issues/263**

## Changes proposed in this Pull Request
- Implementing feature to cluster by NUTS shapes (NUTS0 to NUTS3 for EU countries and OSM administration levels 0-1 for BA, MD, UA, and XK)
- Building on PR https://github.com/PyPSA/pypsa-eur/pull/1479 and using updated `nuts3_shapes.geojson`
- Administration levels can be set for entire modelling region or for individual countries

## Example for illustrative purposes
- All European countries set to NUTS0/country level
- DE set to NUTS3
- FR set to NUTS2
- EE set to NUTS0/country_level

![image](https://github.com/user-attachments/assets/2e832bda-e811-4df3-b8e6-c8eac1ff398a)

## Example config
```
clustering:
  mode: admin 
  admin:
    level: 1
    DE: 3 # optional
    FR: 2 # optional
    ES: 0 # optional
  ...
```

## Checklist

- [X] I tested my contribution locally and it works as intended.
- [X] Code and workflow changes are sufficiently documented.
- [X] Changes in configuration options are added in `config/config.default.yaml`.
- [X] Changes in configuration options are documented in `doc/configtables/*.csv`.
- [X] Sources of newly added data are documented in `doc/data_sources.rst` (Done here: https://github.com/PyPSA/pypsa-eur/pull/1479)
- [X] A release note `doc/release_notes.rst` is added.

FabianHofmann commented on scripts/cluster_network.py:
```diff
@@ -338,8 +436,13 @@ def cluster_regions(
 
-        custom_busmap = params.custom_busmap
-        if custom_busmap:
+        if mode == "administrative":
```
I guess, this too should be == "adm" ?

bobbyxng replied:
"adm" is only for the cluster wildcard to keep the file names short. But in theory we can use the "adm" string everywhere, replacing "administrative" for consistency. My thoughts where short file names vs. "administrative" being clearer/easier to understand for the user, now it's mixed. All three options are fine though :)

FabianHofmann reviewed (commented):
Just a heads-up: In two recent projects I was working on a similar implementation but for clustering on bidding zone level. In near future I would propose this to pypsa-eur as well, to avoid redundancy I would then try to consolidate the functions. My logic was to implement a function `busmap_from_custom_shapes` (or similar) which takes the shapes defined by the user into accound and creates a busmap from there.

bobbyxng commented:
> Just a heads-up: In two recent projects I was working on a similar implementation but for clustering on bidding zone level. In near future I would propose this to pypsa-eur as well, to avoid redundancy I would then try to consolidate the functions. My logic was to implement a function `busmap_from_custom_shapes` (or similar) which takes the shapes defined by the user into accound and creates a busmap from there.

Yes, thanks for pointing that out @FabianHofmann ! We actually had the exact same thought, but wanted to keep this PR small and add additional functionalities later. One could for example also use NUTS3 regions as the building blocks for bidding zones (similiar to the custom busmap logic)

fneum reviewed (changes requested):
Theoretical review first. I'll do a few practical tests now.

bobbyxng commented on doc/configtables/clustering.csv:
```diff
@@ -1,5 +1,9 @@
+mode,str,"One of {‘busmap’, ‘custom_busmap’, ‘administrative’}","‘busmap’: Default. ‘custom_busmap’: Enable the use of custom busmaps in rule mod:`cluster_network`. If activated the rule looks for provided busmaps at ``data/busmaps/base_s_{clusters}_{base_network}.csv`` which should have the same format as ``resources/busmap_base_s_{clusters}.csv``, i.e. the index should contain the buses of ``ne ...
+focus_weights,,,"Optionally specify the focus weights for the clustering of countries. For instance: `DE: 0.8` will distribute 80% of all nodes to Germany and 20% to the rest of the countries. Only applies when mode is set to `busmap`."
+administrative,,,
+-- level,int,"{0, 1, 2, 3}","Level of administrative regions to cluster the network. 0: Country level, 1: NUTS1 level, 2: NUTS2 level, 3: NUTS3 level. Only applies when mode is set to `administrative`. Note that non-NUTS countries 'BA', 'MD', 'UA', and 'XK' can only be clustered to level 0 and 1."
```
it will automatically always fall back to the highest available level. The "trick" is that the nuts_3_shapes.geojson file contains administration level 1 data in level 2 and 3 columns

bobbyxng commented on config/config.default.yaml:
```diff
@@ -865,6 +864,9 @@ costs:
 
 # docs in https://pypsa-eur.readthedocs.io/en/latest/configuration.html#clustering
 clustering:
+  mode: busmap # busmap, custom_busmap, administrative
+  administrative:
+    level: 1 # Set general level of administrative division, e.g. 0 for countries, 1 for NUTS1, etc. Optionally specify on a per-country basis using dict.
```
removed!

bobbyxng commented on doc/release_notes.rst:
```diff
@@ -274,6 +278,8 @@ PyPSA-Eur v2025.01.0 (24th January 2025)
   no changes are needed. But we recommend to set it up locally as well.
   (https://github.com/PyPSA/pypsa-eur/pull/1474)
 
+* Updating all base shapes (country_shapes, europe_shape, nuts3_shapes, ...). The workflow has been modified to use higher resolution and more harmonised shapes (NUTS3 2021 01M data and OSM administration level 1 for non-NUTS3 countries, such as BA, MD, UA, and XK). Data sources for population and GDP p.c. have been updated to JRC ARDECO https://urban.jrc.ec.europa.eu/ardeco/ -- 2019 values are u ...
+
```
ok!

bobbyxng commented on scripts/cluster_network.py:
```diff
@@ -327,6 +423,8 @@ def cluster_regions(
 
     if snakemake.wildcards.clusters == "all":
         n_clusters = len(n.buses)
+    elif snakemake.wildcards.clusters == "adm":
```
right! corrected

bobbyxng commented:
Thanks for the comments @fneum !
- I have further improved the workflow by outsourcing the regionalisation to base_network so that all subsequent shapes in resources (`regions_onshore_XX`) also respect the NUTS boundaries. Voronoi cells are now only formed within administrative regions, if administrative clustering is selected.
- `cluster_network.py` only makes use of the newly generated `admin_shapes.geojson` to map the buses. In case no administrative clustering is selected, regionalisation in `admin_shapes.geojson` will default to administration level 0 / country level.

![image](https://github.com/user-attachments/assets/60c9bfaf-86a2-41e4-8d00-caf8fcc8a204)

bobbyxng commented:
Open edge case (work-in-progress):
At high resolutions (such as NUTS3), "holes" can form, meaning areas without a substation (see yellow spots in screenshot, reds are substations)
![image](https://github.com/user-attachments/assets/43610a8c-609e-41a2-9bf1-e4db7250a50c)

- [ ] For these cases, proposal to merge the NUTS3/NUTSX region with a neighbour of the same NUTS2/NUTS(X-1). If multiple neighbours exist, pick the neighbour with the smallest area.

bobbyxng commented:
**Update**
I have finalised the implementation and added the following:
- For regions that do not contain a substation, implemented an algorithm: Find all neighbours of the same parent region, if no neighbour exist, allow for neighbours of other parent region (however, same country), find the neighbour with the most substations, then sort by smallest area (ascending). --> Merge with neighbour that has most substations and is the smallest in size (secondary). Algorithm is recursive. If all countries are set to level 3, this takes about 30 seconds to finish. 
- Any combination of admin levels (country, admin level 1-3) is technically possible
- `cluster_network`: x, y coordinates are updated to internal points (Pole of Inaccessibility), so the clustered node is guaranteed to be within the largest polygon of each clustered administrative region
- London area is always clustered to level 1 (GBI), as I have observed converging issues in the past

**Screenshots**
- Updated x, y coordinates (yellow before/centroid; red new/PoI):
![Screenshot from 2025-03-07 11-48-27](https://github.com/user-attachments/assets/13294303-68e1-4632-81a5-e100cf1353c0)
![Screenshot from 2025-03-07 11-49-26](https://github.com/user-attachments/assets/49a327e7-abd5-4e6c-9484-3bcedc060d16)

- Aggregated, merged admin regions without "holes".
- In this example, all countries are set to level 1, GR to level 2, DK, DE, GB to level 3
![image](https://github.com/user-attachments/assets/371dc2ce-56fe-4fdf-b123-a5e2fd9e523d)

fneum reviewed (approved):
LGTM - Thanks a lot @bobbyxng for taking care of it!
